### ISSUE TITLE
twisted.words.test.test_jabberclient.IQAuthInitializerTest.testDigest fails on Python trunk@HEAD

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3741]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3741|
|Type|defect|
|Created|2009-03-31 21:15:53Z|

```
===============================================================================
[FAIL]: twisted.words.test.test_jabberclient.IQAuthInitializerTest.testDigest

Traceback (most recent call last):
  File "/home/exarkun/Projects/Twisted/trunk/twisted/words/test/test_jabberclient.py", line 116, in testDigest
    self.assertEquals('set', iq['type'])
twisted.trial.unittest.FailTest: not equal:
a = 'set'
b = 'result'

```

<details><summary>Searchable metadata</summary>

```
trac-id__3741 3741
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__Python_2_7 Python-2.7
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1238534153000000 1238534153000000
changetime__1261996730000000 1261996730000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ivank commented|
|-|-|

I am pretty sure that the patch in [#3847](https://github.com/twisted/twisted/issues/3847) fixes this, too.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm set status to `assigned`|
|-|-|

Fixing [[#3847](https://github.com/twisted/twisted/issues/3847)](https://github.com/twisted/twisted/issues/3847) does indeed obscure the actual bug in this test again.

The problem is really that the test looks at the last sent stanza that was generated by the initializer twice. The first part of the text modifies the sent request stanza (for retrieving the authentication parameters) to create a response, changing the stanza type to 'result'.

Unfortunately, because #3847 raises an exception, the next stanza, the actual request to authenticate, never gets sent, so `self.output[-1]` now contains the response we just sent out.

The test should be modified to:

 * Not reuse the items in `self.output` to create a response, but instead use `t.w.p.j.xmlstream.toResponse`.
 * Detect that there has been no further output, instead of looking at the wrong stanza.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm commented|
|-|-|

(In [27713]) Make sure input is encoded to UTF-8 before calculating hash.

Addresses [#3741](https://github.com/twisted/twisted/issues/3741), [#3742](https://github.com/twisted/twisted/issues/3742), [#3847](https://github.com/twisted/twisted/issues/3847).

The fix for `hashPassword` was rather easy, but highlights issues in how the
tests in `test_jabberxmlstream` work.

These tests work for the happy case, but exceptions can be raised
asynchronously in XPath observers, causing spurious errors in unrelated tests
([#3847](https://github.com/twisted/twisted/issues/3847)) if the actual test also raises an exception causing the deferred to not
be returned.

I addressed this by rewriting all of those tests so the handlers also trigger
deferreds we can wait for.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm commented|
|-|-|

(In [27772]) Explicitly test for unicode as input for hashPassword, add __all__.

Re [#3741](https://github.com/twisted/twisted/issues/3741), [#3742](https://github.com/twisted/twisted/issues/3742), [#3847](https://github.com/twisted/twisted/issues/3847).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set status to `closed`|
|-|-|

(In [27896]) Fix failing tests for Python 2.7, allow non-ascii secrets in hashPassword.

Author: ralphm.
Reviewer: exarkun, TimAllen.
Fixes [#3741](https://github.com/twisted/twisted/issues/3741), [#3742](https://github.com/twisted/twisted/issues/3742), [#3847](https://github.com/twisted/twisted/issues/3847).

This lets twisted.words.protocols.jabber.xmlstream.hashPassword now only accept
unicode as input. The input before hashing is encoded to UTF-8. Tests for
hashPassword were added and a number of tests were rewritten to deal with
asynchronous failures.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
